src/Sulu/Bundle/AdminBundle/Resources/js/components/Tabs/Tabs.js

```diff
@@ -1,56 +1,398 @@
+        this.dropdownOpen = !this.dropdownOpen;
+    };
+
+    @action handleDropdownClose = () => {
+        this.dropdownOpen = false;
+    };
+
+    changeTab = (selectedTabIndex: ?number) => {
+        if (selectedTabIndex !== undefined && selectedTabIndex !== null) {
+            this.props.onSelect(selectedTabIndex);
+        }
+    };
+
+    handleTabClick = (selectedTabIndex: ?number) => {
+        this.changeTab(selectedTabIndex);
+    };
+
+    @action handleCollapsedTabClick = (selectedTabIndex: number) => {
+        this.dropdownOpen = false;
+        this.lastSelectedIndex = selectedTabIndex;
+
+        this.changeTab(selectedTabIndex);
+    };
+
     isSelected(tabIndex: number) {
         return tabIndex === this.props.selectedIndex;
     }
 
-    createTabItems(tabs: ChildrenArray<Element<typeof Tab> | false>) {
-        return React.Children.map(tabs, (tab, index) => {
+    get childIndices(): number[] {
+        const {children} = this.props;
+
+        return React.Children.map(children, (child, index) => index);
+    }
+
+    @computed get visibleTabIndices(): number[] {
+        if (this.tabsWidth <= this.wrapperWidth) {
+            return this.childIndices;
+        }
+
+        const {selectedIndex} = this.props;
+
+        let visibleWidth = 0;
+        let visibleTabIndices: number[] = [];
+
+        const childIndices = this.childIndices;
+        childIndices.sort((a, b) => {
+            if (a === selectedIndex) {
+                return -1;
+            }
+
+            if (b === selectedIndex) {
+                return 1;
+            }
+
+            if (a === this.lastSelectedIndex) {
+                return -1;
+            }
+
+            if (b === this.lastSelectedIndex) {
+                return 1;
+            }
+
+            return a - b;
+        });
+
+        for (const index of childIndices) {
+            const nextWidth = this.tabWidths.get(index);
+
+            if (undefined === nextWidth) {
+                if (visibleTabIndices.length > 0) {
+                    break;
+                }
+
+                return this.childIndices;
+            }
+
+            if (visibleWidth + nextWidth > this.tabsWrapperWidth) {
+                break;
+            }
+
+            visibleWidth += nextWidth;
+            visibleTabIndices = [...visibleTabIndices, index];
+        }
+
+        visibleTabIndices.sort((a, b) => a - b);
+
+        return visibleTabIndices;
+    }
+
+    @computed get collapsedTabIndices(): number[] {
+        const visibleTabIndices = this.visibleTabIndices;
+
+        return this.childIndices.filter((index) => !visibleTabIndices.includes(index));
+    }
+
+    @computed get hasCollapsedTabs(): boolean {
+        return this.collapsedTabIndices.length > 0;
+    }
+
+    get tabs() {
+        const {children} = this.props;
+        const visibleTabIndices = this.visibleTabIndices;
+        const collapsedTabIndices = this.collapsedTabIndices;
```

Comment:
The previous two lines are not really necessary, are they? And if you just want to write less, destructuring would save even more :slightly_smiling_face:

Reply:
I use `visibleTabIndices` and `collapsedTabIndices` at two places each and I'm not sure if calling using `this.visibleTabIndices` twice would result in two calls to `get visibleTabIndices() {}`. But yes, using destructuring would indeed be better 🙂